partner_affiliate/models/res_partner.py

```diff
@@ -23,3 +23,31 @@ class ResPartner(models.Model):
                                     string='Affiliates',
                                     domain=[('active', '=', True),
                                             ('is_company', '=', True)])
+
+    @api.onchange('parent_id')
+    def onchange_parent_id(self):
+        # Keep the original address info to set it back if is a company.
+        original_address = self.get_original_address()
+
+        new_partner = super(ResPartner, self).onchange_parent_id()
+
+        # When the affiliate is a company, we must set back its address
+        # because the super call changes its address by the new parent address.
+        # In addition, the type must be set to affiliate instead of contact.
+        if self.is_company:
+            self.type = 'affiliate'
+            self.update_address(original_address)
+            return
```

Comment:
super(ResPartner, self).onchange_parent_id() will return some values.
Here you have to return that to keep the inheritance mechanism, but with the fields updated. 
You cannot do NEVER a write from an onchange, as maybe the user is still creating the partner and the onchange is called. Then you should not write in a partner that is still not created. 

You have to refactor this two lines, so that you return the value instead of doing a write

self.type = 'affiliate'
self.update_address(original_address)

Instead use something like 

res = super(ResPartner, self).onchange_parent_id()

res['value'].update(original_address)
res['value'].update({'type': 'affiliate'})
return res

Reply:
Ok, changed